# IBM Coursera Applied Data Science Capstone Project Week 3

## Segmenting and Clustering Neighborhoods in Toronto

In [1]:
# import libraries
import pandas as pd
import numpy as np

# Part 1: Wikipedia page data transformed into pandas dataframe

Code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe

In [2]:
# define source url and read data to dataframe. Note that the wikipedia data table has changed, 
# so one needs to refer to and older version of the website

url = 'https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&oldid=890001620'
df = pd.read_html(url, header=None)[0]
df.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [3]:
# drop all rows where Borough is "Not assigned"
df = df[df.Borough != 'Not assigned']

# change column names to match the names given in the assignment
df.columns = ['PostalCode', 'Borough', 'Neighborhood']

# reset index
df = df.reset_index(drop=True)
df.head()


PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M5A  Downtown Toronto       Regent Park
4        M6A        North York  Lawrence Heights

In [4]:
# Neighborhoods with same postal code area will be combined into one row with the neighborhoods 
# separated with a comma

df_grouped = df.groupby(['Borough', 'PostalCode'])['Neighborhood'].apply(', '.join).reset_index()
df_grouped.head()

Borough PostalCode                 Neighborhood
0  Central Toronto        M4N                Lawrence Park
1  Central Toronto        M4P             Davisville North
2  Central Toronto        M4R           North Toronto West
3  Central Toronto        M4S                   Davisville
4  Central Toronto        M4T  Moore Park, Summerhill East

In [5]:
# rearrange columns

cols = ['PostalCode', 'Borough', 'Neighborhood']
df_grouped = df_grouped[cols]
df_grouped.head(12)

PostalCode           Borough  \
0         M4N   Central Toronto   
1         M4P   Central Toronto   
2         M4R   Central Toronto   
3         M4S   Central Toronto   
4         M4T   Central Toronto   
5         M4V   Central Toronto   
6         M5N   Central Toronto   
7         M5P   Central Toronto   
8         M5R   Central Toronto   
9         M4W  Downtown Toronto   
10        M4X  Downtown Toronto   
11        M4Y  Downtown Toronto   

                                         Neighborhood  
0                                       Lawrence Park  
1                                    Davisville North  
2                                  North Toronto West  
3                                          Davisville  
4                         Moore Park, Summerhill East  
5   Deer Park, Forest Hill SE, Rathnelly, South Hi...  
6                                            Roselawn  
7                 Forest Hill North, Forest Hill West  
8                 The Annex, North Midtown, Yorkville  
9                                            Rosedale  
10                        Cabbagetown, St. James Town  
11                               Church and Wellesley

In [6]:
#  check if Neighborhoods for Postal Code M5A matches Harbourfront and Regent Park

print (df_grouped.loc[df_grouped['PostalCode'] == 'M5A'])

   PostalCode           Borough               Neighborhood
12        M5A  Downtown Toronto  Harbourfront, Regent Park


In [7]:
# show shape as requested by the assignment

df_grouped.shape

(103, 3)

# Part 2: Dataframe with geographical coordinates


Next, I want to get the latitude and the longitude coordinates of each neighborhood. First, I use the Geocoder Python package as described in the assignment.


In [8]:
!pip install geocoder

In [9]:
import geocoder # import geocoder

Unfortunately, when I run the following code it takes ages to retrieve the data. That's why I skipped this task and move on the work with the csv-file.

def get_latlng(postal_code):
    # initialize your variable to None
    lat_lng_coords = None

    # loop until you get the coordinates
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
    return lat_lng_coords

get_latlng('M4G')
#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

To make sure that the data I will use in the next part of the assignment is correct, I will now retrieve the coordinates from the csv file.  

In [10]:
# define source url
url = 'https://cocl.us/Geospatial_data'

# create new dataframe with coordinates
df_geo = pd.read_csv(url)

# modify the column names
df_geo.columns = ['PostalCode', 'Latitude', 'Longitude']
df_geo.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [11]:
# merge both tables, from part 1 with the neighborhood data and from part 2 with the coordinates

df = df_grouped.merge(df_geo, on='PostalCode', how='left')
df.head()

PostalCode          Borough                 Neighborhood   Latitude  \
0        M4N  Central Toronto                Lawrence Park  43.728020   
1        M4P  Central Toronto             Davisville North  43.712751   
2        M4R  Central Toronto           North Toronto West  43.715383   
3        M4S  Central Toronto                   Davisville  43.704324   
4        M4T  Central Toronto  Moore Park, Summerhill East  43.689574   

   Longitude  
0 -79.388790  
1 -79.390197  
2 -79.405678  
3 -79.388790  
4 -79.383160

In [12]:
# Check table size

df.shape

(103, 5)










# Part 3: Explore and cluster the neighborhoods in Toronto

In part 3 I will explore and cluster the neighborhoods in Toronto. I decided to work only with boroughs that contain the word Toronto. What follows is more or less a replication of the same analysis that was done in the labs with New York City data.


Before we start exploring the data, let's download all the dependencies that we will need.

In [13]:
!pip install geopy

In [14]:
!pip install folium

In [15]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy=1.19.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge --no-deps altair --yes
!conda install -c conda-forge --no-deps vincent --yes

!conda install -c conda-forge --no-deps folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

usage: conda-script.py [-h] [-V] command ...
conda-script.py: error: unrecognized arguments: # uncomment this line if you haven't completed the Foursquare API lab


Solving environment: ...working... done

# All requested packages already installed.

Solving environment: ...working... done

# All requested packages already installed.



usage: conda-script.py [-h] [-V] command ...
conda-script.py: error: unrecognized arguments: # uncomment this line if you haven't completed the Foursquare API lab


Libraries imported.


#### Load and explore the data

We build on the dataframe created above to explore the neighborhoods in Toronto. 

In [16]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df['Borough'].unique()),
        df.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


The dataframe has a total of 11 boroughs and 103 neighborhoods. 

In order to segement the neighborhoods and explore them, we will need a dataset that contains the 11 boroughs and the neighborhoods that exist in each borough as well as the the latitude and logitude coordinates of each neighborhood.

That task is already finished in Part 1 and 2 of this assignment.

#### Use geopy library to get the latitude and longitude values of Toronto.

In order to define an instance of the geocoder, I define a user_agent. The name of the agent is *toro_explorer*, as shown below.

In [17]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="toro_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top

In [18]:
# create a new dataframe based on part 2 and the same name as in the labs

neighborhoods = df
neighborhoods.head()

PostalCode          Borough                 Neighborhood   Latitude  \
0        M4N  Central Toronto                Lawrence Park  43.728020   
1        M4P  Central Toronto             Davisville North  43.712751   
2        M4R  Central Toronto           North Toronto West  43.715383   
3        M4S  Central Toronto                   Davisville  43.704324   
4        M4T  Central Toronto  Moore Park, Summerhill East  43.689574   

   Longitude  
0 -79.388790  
1 -79.390197  
2 -79.405678  
3 -79.388790  
4 -79.383160

In [19]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Now I will split the dataframe into two sets and continue working with the boroughs that contain the word "Toronto".  

In [20]:
# create new dataframe with "Toronto"-borouhgs

df_toro = neighborhoods[neighborhoods['Borough'].str.contains("Toronto")].reset_index(drop=True)
df_toro.head()

PostalCode          Borough                 Neighborhood   Latitude  \
0        M4N  Central Toronto                Lawrence Park  43.728020   
1        M4P  Central Toronto             Davisville North  43.712751   
2        M4R  Central Toronto           North Toronto West  43.715383   
3        M4S  Central Toronto                   Davisville  43.704324   
4        M4T  Central Toronto  Moore Park, Summerhill East  43.689574   

   Longitude  
0 -79.388790  
1 -79.390197  
2 -79.405678  
3 -79.388790  
4 -79.383160

In [21]:
# Let's check what's left from the big dataframe with all boroughs

df_toro.shape

(38, 5)

In [22]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_toro['Borough'].unique()),
        df_toro.shape[0]
    )
)

The dataframe has 4 boroughs and 38 neighborhoods.


In [23]:
# create another map of Toronto using latitude and longitude values, but this time only with inner borouhgs
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toro['Latitude'], df_toro['Longitude'], df_toro['Borough'], df_toro['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Now we're down from 11 boroughs and 103 neighborhoods to 4 boroughs and 38 neighborhoods.


Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [24]:
CLIENT_ID = 'MKBS1TQLR2D0RUWF4APKGKYC5AHKWQ5GMC0JECF5P0TMUUWL' # your Foursquare ID
CLIENT_SECRET = '1UYPQ0RXCN3FOR1DMMR4MH3B0ZJIVWG0ANA5KRTUGRTWAK20' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

LIMIT = 100
radius = 500

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: MKBS1TQLR2D0RUWF4APKGKYC5AHKWQ5GMC0JECF5P0TMUUWL
CLIENT_SECRET:1UYPQ0RXCN3FOR1DMMR4MH3B0ZJIVWG0ANA5KRTUGRTWAK20


#### Let's explore neighborhoods in Inner-Toronto

Let's create a function that does the following: 
- gets the top 100 venues within a radius of 500 meters
- cleans the json file and structures it into a pandas dataframe

The function does the above to all the neighborhoods in our sample of Toronto boroughs. 

In [25]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

toronto_venues = getNearbyVenues(names=df_toro['Neighborhood'],
                                   latitudes=df_toro['Latitude'],
                                   longitudes=df_toro['Longitude']
                                  )


Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Business Reply Mail Processing Centre 969 Eastern
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, 

#### Let's check the size of the resulting dataframe

In [26]:
print(toronto_venues.shape)
toronto_venues.head()

(1649, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     Lawrence Park              43.728020              -79.388790   
1     Lawrence Park              43.728020              -79.388790   
2     Lawrence Park              43.728020              -79.388790   
3     Lawrence Park              43.728020              -79.388790   
4  Davisville North              43.712751              -79.390197   

                                          Venue  Venue Latitude  \
0                          Lawrence Park Ravine       43.726963   
1  J.N. Munkongolo & Associates - Family Lawyer       43.727437   
2                            Zodiac Swim School       43.728532   
3                TTC Bus #162 - Lawrence-Donway       43.728026   
4                       Summerhill Market North       43.715499   

   Venue Longitude     Venue Category  
0       -79.394382               Park  
1       -79.390264             Lawyer  
2       -79.382860        Swim School  
3       -79.382805           Bus Line  
4       -79.392881  Food & Drink Shop

Let's check how many venues were returned for each neighborhood

In [27]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            55   
Brockton, Exhibition Place, Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      16   
CN Tower, Bathurst Quay, Island airport, Harbou...                     16   
Cabbagetown, St. James Town                                            47   
Central Bay Street                                                     77   
Chinatown, Grange Park, Kensington Market                              76   
Christie                                                               19   
Church and Wellesley                                                   79   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     15   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           16   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         35   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              45   
High Park, The Junction South                                          24   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               50   
Moore Park, Summerhill East                                             2   
North Toronto West                                                     17   
Parkdale, Roncesvalles                                                 13   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     40   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        95   
Studio District                                                        42   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             5   
The Beaches West, India Bazaar                                         21   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             55   
Brockton, Exhibition Place, Parkdale Village                            22   
Business Reply Mail Processing Centre 969 Eastern                       16   
CN Tower, Bathurst Quay, Island airport, Harbou...                      16   
Cabbagetown, St. James Town                                             47   
Central Bay Street                                                      77   
Chinatown, Grange Park, Kensington Market                               76   
Christie                                                                19   
Church and Wellesley                                                    79   
Commerce Court, Victoria Hotel     

Let's find out how many unique categories can be curated from all the returned venues¶

In [28]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 234 uniques categories.


### Analyze Each Neighborhood

In [29]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         1   
4        0                0         0             0              0         0   

   Business Service  Butcher  Café  Cajun / Creole Restaurant  Camera Store  \
0                 0        0     0                          0             0   
1                 0        0     0                          0             0   
2                 0        0     0                          0             0   
3

And let's examine the new dataframe size.

In [30]:
toronto_onehot.shape

(1649, 234)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category¶

In [31]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.062500   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.012987   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.025316   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.028571   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.022222   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.020000   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.058824   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.025000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.023810   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023810   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0000              0.0000        0.0000   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0000              0.0000        0.0000   
4            0.000000   0.0625              0.0625        0.0625   
5            0.000000   0.0000              0.0000        0.0000   
6            0.000000   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.012658   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.000000   0.0000              0.0000        0.0000   
18           0.000

Let's confirm the new size

In [32]:
toronto_grouped.shape

(38, 234)

### Let's print each neighborhood along with the top 5 most common venues

In [33]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
         venue  freq
0  Coffee Shop  0.06
1   Restaurant  0.06
2         Café  0.06
3        Hotel  0.03
4          Bar  0.03


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2            Beer Bar  0.04
3         Cheese Shop  0.04
4  Seafood Restaurant  0.04


----Brockton, Exhibition Place, Parkdale Village----
                venue  freq
0                Café  0.14
1         Coffee Shop  0.09
2      Breakfast Spot  0.09
3                 Gym  0.05
4  Italian Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
           venue  freq
0    Yoga Studio  0.06
1  Auto Workshop  0.06
2           Park  0.06
3     Comic Shop  0.06
4    Pizza Place  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3       C

Let's put that into a pandas dataframe. First, let's write a function to sort the venues in descending order.


In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [35]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...      Airport Terminal   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Restaurant                  Café                   Gym   
1          Cocktail Bar        Farmers Market              Beer Bar   
2           Coffee Shop        Breakfast Spot             Pet Store   
3         Auto Workshop            Comic Shop                  Park   
4        Airport Lounge       Airport Service   Rental Car Location   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery                   Bar                 Hotel   
1                Bakery    Seafood Restaurant            Restaurant   
2               Stadium         Burrito Place            Restaurant   
3           Pizza Place         Burrito Place            Restaurant   
4         Boat or Ferry           Coffee Shop              Boutique   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0       Thai Restaurant                 Office                 Lounge  
1                  Café            Cheese Shop              Nightclub  
2          Climbing Gym  Performing Arts Venue                 Bakery  
3               Brewery     Light Rail Station             Smoke Shop  
4       Harbor / Marina                    Bar           Airport Gate

## Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [36]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [37]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toro

# merge toronto_grouped with df_toro to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() 

PostalCode          Borough                 Neighborhood   Latitude  \
0        M4N  Central Toronto                Lawrence Park  43.728020   
1        M4P  Central Toronto             Davisville North  43.712751   
2        M4R  Central Toronto           North Toronto West  43.715383   
3        M4S  Central Toronto                   Davisville  43.704324   
4        M4T  Central Toronto  Moore Park, Summerhill East  43.689574   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.388790               2                Lawyer                  Park   
1 -79.390197               4                   Gym     Convenience Store   
2 -79.405678               0        Clothing Store           Coffee Shop   
3 -79.388790               0        Sandwich Place          Dessert Shop   
4 -79.383160               3            Playground           Summer Camp   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0              Bus Line           Swim School                 Diner   
1        Breakfast Spot        Sandwich Place     Food & Drink Shop   
2   Sporting Goods Shop  Fast Food Restaurant                 Diner   
3           Pizza Place                   Gym           Coffee Shop   
4         Women's Store          Dessert Shop           Event Space   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0           Event Space  Ethiopian Restaurant            Electronics Store   
1                 Hotel      Department Store                         Park   
2          Dessert Shop    Mexican Restaurant                         Park   
3    Italian Restaurant                  Café             Sushi Restaurant   
4  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

         9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant    Dumpling Restaurant  
1               Cosmetics Shop        Coworking Space  
2           Chinese Restaurant                   Café  
3                         Park             Restaurant  
4          Dumpling Restaurant             Donut Shop

Finally, let's visualize the resulting clusters

In [38]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## That's it. Thank you for reviewing my assignment!